replication_test.go

```diff
@@ -0,0 +1,187 @@
+	laggingNode := newSimplexNode(t, nodes[3], net, bb)
+
+	require.Equal(t, startSeq, normalNode1.storage.Height())
+	require.Equal(t, startSeq, normalNode2.storage.Height())
+	require.Equal(t, startSeq, normalNode3.storage.Height())
+	require.Equal(t, uint64(0), laggingNode.storage.Height())
+
+	net.startInstances()
+	bb.triggerNewBlock()
+
+	// all blocks except the lagging node start at round 8, seq 8.
+	// lagging node starts at round 0, seq 0.
+	// this asserts that the lagging node catches up to the latest round
+	for _, n := range net.instances {
+		n.storage.waitForBlockCommit(uint64(startSeq))
+	}
+}
+
+// TestReplicationExceedsMaxRoundWindow tests the replication process of a node that
+// is behind the rest of the network by more than maxRoundWindow.
+func TestReplicationExceedsMaxRoundWindow(t *testing.T) {
+	bb := newTestControlledBlockBuilder(t)
+	nodes := []simplex.NodeID{{1}, {2}, {3}, []byte("lagging")}
+	net := newInMemNetwork(t, nodes)
+	startSeq := uint64(simplex.DefaultMaxRoundWindow * 3)
+
+	storageData := createBlocks(t, nodes, &bb.testBlockBuilder, startSeq)
+	normalNode1 := newSimplexNodeWithStorage(t, nodes[0], net, bb, storageData)
+	normalNode2 := newSimplexNodeWithStorage(t, nodes[1], net, bb, storageData)
+	normalNode3 := newSimplexNodeWithStorage(t, nodes[2], net, bb, storageData)
+	laggingNode := newSimplexNode(t, nodes[3], net, bb)
+	require.Equal(t, startSeq, normalNode1.storage.Height())
+	require.Equal(t, startSeq, normalNode2.storage.Height())
+	require.Equal(t, startSeq, normalNode3.storage.Height())
+	require.Equal(t, uint64(0), laggingNode.storage.Height())
+
+	net.startInstances()
+	bb.triggerNewBlock()
+	for _, n := range net.instances {
+		n.storage.waitForBlockCommit(uint64(startSeq))
+	}
+}
+
+// TestReplicationStartsBeforeCurrentRound tests the replication process of a node that
+// starts replicating in the middle of the current round.
+func TestReplicationStartsBeforeCurrentRound(t *testing.T) {
+	bb := newTestControlledBlockBuilder(t)
+	nodes := []simplex.NodeID{{1}, {2}, {3}, []byte("lagging")}
+	quorum := simplex.Quorum(len(nodes))
+	net := newInMemNetwork(t, nodes)
+	startSeq := uint64(simplex.DefaultMaxRoundWindow + 3)
+	storageData := createBlocks(t, nodes, &bb.testBlockBuilder, startSeq)
+
+	normalNode1 := newSimplexNodeWithStorage(t, nodes[0], net, bb, storageData)
+	normalNode2 := newSimplexNodeWithStorage(t, nodes[1], net, bb, storageData)
+	normalNode3 := newSimplexNodeWithStorage(t, nodes[2], net, bb, storageData)
+	laggingNode := newSimplexNode(t, nodes[3], net, bb)
+
+	firstBlock := storageData[0].Block
+	record := simplex.BlockRecord(firstBlock.BlockHeader(), firstBlock.Bytes())
+	laggingNode.wal.Append(record)
+
+	firstNotarizationRecord, err := newNotarizationRecord(laggingNode.e.Logger, laggingNode.e.SignatureAggregator, firstBlock, nodes[0:quorum])
+	require.NoError(t, err)
+	laggingNode.wal.Append(firstNotarizationRecord)
```

Comment:
I think it's better to inject votes / notarizations to the node directly and not rely on recovery logic. We're should aim to replicate (not pun intended) conditions that are as close to what would happen in a real scenario. 

In a real scenario we most likely wouldn't be replicating right after a crash, but while we're up and running.